### DISCUSSION TITLE
Changing bind to OptionalBind instead

### DISCUSSION THREAD

**Discussion by asger-weirsoee** [OxalisCommunity/Oxalis-AS4#193] [General]
I'm working on a team that is actively working on implementing a plugin to Oxalis-AS4 (which we know is a plugin) to allow for another CA chain than PEPPOL. This would require that we were able to change parts of the call chain of the Oxalis-AS4 classes.

Now what we have to do is to check if a specific standard is used if so, we'd execute our specific code, or else we'd the Oxalis-AS4 plugin. 

I've created somewhat of a POC but it would require that at certain places we'd have to replace the GUICE bind structure to use OptionalBind instead, and then place a default value, and I wanted to pick your brains about that.
Currently, we are not 100% certain on how or where we are going to implement our implementation, else I would've created a PR and taken the discussion there instead.

So now instead I thought we could have the discussion in the tab meant for it instead;

One of the place I think we'd want to use OptionalBind instead of bind is in the As4InboundModule.java

```java
public class As4InboundModule extends ServletModule {

    @Override
    protected void configureServlets() {
        bind(AbstractEndpointSelectionInterceptor.class).to(As4EndpointSelector.class);

        bind(Key.get(HttpServlet.class, Names.named("oxalis-as4")))
                .to(As4Servlet.class)
                .asEagerSingleton();

        bind(As4Provider.class);
        OptionalBinder.newOptionalBinder(binder(), As4EndpointsPublisher.class)
                .setDefault()
                .to(As4EndpointsPublisherImpl.class);
        bind(As4InboundHandler.class);

        serve("/as4").with(Key.get(HttpServlet.class, Names.named("oxalis-as4")));
        serve("/as4/status").with(AS4StatusServlet.class);
    }
}
```

This makes it so that in my extension of the AS4 plugin, I can write:

```java
public class DemoInboundModule extends ServletModule {
    @Override
    protected void configureServlets() {

      OptionalBinder.newOptionalBinder(
                binder(), As4EndpointsPublisher.class)
                .setBinding()
                .to(DemoEndpointsPublisherImpl.class);

        bind(Key.get(HttpServlet.class, Names.named("oxalis-demo")))
                .to(DemoServlet.class)
                .asEagerSingleton();
    }
}
```

And thus change which implementation that is bound to the As4EndpointsPublisher interface.

Now what I want to talk with you folks about, is;
Is there a reason for why you've used bind as the method in use, and would it be okay for you, if I created a PR once we found which place we wanted to change your AS4 plugin, to instead utilize the OptionalBind?

**Comment by dladlk** [comment]
Hello Asger, 

just to be sure - if you want only different CA chain, did you evaluate possibility to use another cert-validator ValidatorRecipe XML, used by Oxalis to detect Mode (TEST/PROD/LOCAL) - defined in **peppol-security-2.1.0.jar** **pki** folder and configured there in reference.xml? 

You can prepare similar XML with your rules and CA/Intermediate certs and "overwrite" globally the chain e.g. for PRODUCTION mode and provide another truststore.

Or you want the same instance to run both Peppol and other CA chains?

**Reply by asger-weirsoee** [reply]
Hi Dmytro 

We need to support both PEPPOL and another CA chain. And due to some laws in Denmark, when using the other CA chain we have to make extra validations, which is not currently supported by AS4 nor PEPPOL.


This validation is in essence:
In the 4 corner model, C4 should be able to verify that C1 is the sender of a document without having to involve C2 or C3 in an investigative journey (per say).
We are planning on implementing a simplified ASiC-e plugin that our modifications to AS4 should call, when using the other CA chain.
